### PR TITLE
URL Encoding Updates

### PR DESCRIPTION
**Pull request by ben-jones**
Two main pieces of functionality here
1. Can now encode urls so they look they Baidu search queries
2. Bug fixes for market url encoding and changed way that padding is handled

### CODE REVIEW

File: htpt/urlEncode.py [hunk 1]
```diff
@@ -151,18 +155,28 @@ def encodeAsMarket(data):
...
+    padding = []
+    for index in range(padSize):
+      padding.append(pickRandomHexChar())
+    padding = ''.join(padding)
+  else:
+    padSize = 0
+    padding = ''
+  dataLen = hex(len(hexData))[2:]
+  if len(dataLen) == 1:
+    dataLen = '0' + dataLen
```
Comment (gsathya) [inline htpt/urlEncode.py:181]: What does this do?

File: htpt/urlEncode.py [hunk 2]
```diff
@@ -179,6 +193,107 @@ def isMarket(url):
...
+  if len(data) > 40:
+    urlData = data[:40]
+    cookies = encodeAsCookies(data[40:])
+  words = encodeAsEnglish(urlData)
+  urlData = '+'.join(words)
+  #Note: we cannot use urlparse here because it capitalizes our hex
+  #values and we are using uppercase to distinguish padding and
+  #actual text
+  url = 'http://www.baidu.com/s?wd=' + urlData
+  retVal = {'url':url, 'cookie':cookies}
```
Comment (gsathya) [inline htpt/urlEncode.py]: Please use something like `encodedData` instead of `retVal`

File: htpt/urlEncode.py [hunk 3]
```diff
@@ -188,12 +303,12 @@ def decodeAsMarket(url):
   Returns: a string with the decoded data
 
   """
-
   pattern = 'click.*\?qs=(?P<hash>[0-9a-fA-F]*)'
   matches = re.search(pattern, url)
   data = matches.group('hash')
   # strip any padding
-  data = data.strip('ABCDEF')
+  dataLen = int(data[:2], 16)
```
Comment (gsathya) [inline htpt/urlEncode.py:313]: why 16? why 2?

File: htpt/urlEncode.py (hunk 1 above)
Comment (ben-jones) [inline htpt/urlEncode.py:181]: it pads the data length field. If the data is shorter than 16 characters, then the data field only has 1 nibble, and we need the length to always be 2 hex character, aka 1 byte.

File: htpt/urlEncode.py (hunk 3 above)
Comment (ben-jones) [inline htpt/urlEncode.py:313]: This is a one liner to convert the included length info into an integer value that I can use. The 16 tells python what numbering system this is coming from, in this case, it is telling python to interpret the characters as hex.

File: htpt/urlEncode.py (hunk 1 above)
Comment (gsathya) [inline htpt/urlEncode.py:181]: Please add a comment explaining this. It's not terribly obvious

File: htpt/urlEncode.py (hunk 3 above)
Comment (gsathya) [inline htpt/urlEncode.py:313]: Please add a comment explaining this?

File: tests/verifyUrlEncode.py [hunk 4]
```diff
@@ -129,8 +130,9 @@ def test_encodeAsMarket(self):
         #the data
         hashPart = matches.group('hash')
         self.assertEqual(len(hashPart), 80)
-        hashPart = hashPart.rstrip('ABCDEF')
-        testDatum = binascii.unhexlify(hashPart)
+        dataLen = int(hashPart[:2], 16)
```
Comment (gsathya) [inline tests/verifyUrlEncode.py:136]: Please add a comment here as well :)

Comment (gsathya) [inline tests/verifyUrlEncode.py:136]: Or not .. since this is just a test. Clearly I need to sleep soon

**Comment by gsathya** [comment]
Hey @ben-jones, can you add the comments so that I can merge this? Thanks

**Comment by gsathya** [comment]
Merged. Thanks! :)

**Pull request merged**